File: lib/server/secret_store.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:logging/logging.dart';
import 'package:path/path.dart' as p;
import 'env.dart';

final _log = Logger('SecretStore');

enum SecretStoreStatus {
  /// No age private key was supplied at startup. Server runs but services
  /// that need encrypted secrets cannot start (their commands will sops-fail).
  missingKey,

  /// Key supplied + stamp existed + decrypted successfully.
  verified,

  /// Key supplied + no stamp on disk → just created one with this key.
  newStamp,

  /// Key supplied + stamp existed + sops decryption failed. `bin/server.dart`
  /// treats this as fatal and exits.
  mismatch,
}

class SecretStoreLockedException implements Exception {
  final String message;
  SecretStoreLockedException(this.message);
  @override
  String toString() => message;
}

/// In-memory secrets gateway, backed by `sops` + `age`. Holds the dev's age
/// private key (pasted at startup via the wrapper) in memory; never touches
/// disk with it. Public key is derived once and cached.
///
/// All encryption / decryption is delegated to the `sops` CLI — see
/// docker/Dockerfile for the pinned versions. Limousine does NOT inject
/// secrets into spawned services anymore; instead the start command in each
/// project's limousine.proj does `sops exec-env .env.secrets -- …` itself.
/// This module's runtime responsibilities are:
///   - validate the key on startup against a `secret-store-stamp` (sops-encrypted)
///   - serve the secrets editor UI (decrypt for view, encrypt on save)
///   - expose the private key so service_manager can forward SOPS_AGE_KEY
///     into spawned child processes
///
/// Deliberately not called "vault" — that name belongs to HashiCorp Vault,
/// and this is a much smaller, single-key local-disk feature.
class SecretStore {
  String? _privateKey; // AGE-SECRET-KEY-…
  String? _publicKey;  // age1…
  SecretStoreStatus _status = SecretStoreStatus.missingKey;
  String? _workspaceDir;

  static const _stampPlaintext = 'limousine-secret-store-v2';

  final StreamController<SecretStoreStatus> _changes =
      StreamController<SecretStoreStatus>.broadcast();

  Stream<SecretStoreStatus> get changes => _changes.stream;
  SecretStoreStatus get status => _status;
  bool get unlocked => _privateKey != null;

  /// The age private key, exposed so service_manager can forward it into
  /// child processes as SOPS_AGE_KEY. Null if the store is locked.
  String? get privateKey => _privateKey;

  String? stampPathFor(String workspacePath) =>
      p.join(p.dirname(workspacePath), '.limousine', 'secret-store-stamp');

  void _setStatus(SecretStoreStatus s) {
    _status = s;
    _changes.add(s);
  }

  /// Read `LIMOUSINE_AGE_KEY` from the environment, validate it by deriving
  /// the public key, then verify or create the workspace stamp file.
  Future<void> initForWorkspace(String workspacePath) async {
    _workspaceDir = p.dirname(workspacePath);
    final keyEnv = Platform.environment['LIMOUSINE_AGE_KEY'];
    _privateKey = null;
    _publicKey = null;

    if (keyEnv == null || keyEnv.isEmpty) {
      _setStatus(SecretStoreStatus.missingKey);
      _log.warning(
        'LIMOUSINE_AGE_KEY not set — secret store locked. '
        'Services with encrypted secrets will fail to decrypt.',
      );
      return;
    }

    final pub = await _derivePublicKey(keyEnv.trim());
    if (pub == null) {
      _setStatus(SecretStoreStatus.mismatch);
      _log.severe(
        'LIMOUSINE_AGE_KEY does not parse as a valid age private key. '
        'Restart with a key produced by `age-keygen`.',
      );
      return;
    }
    _privateKey = keyEnv.trim();
    _publicKey = pub;

    final stampPath = stampPathFor(workspacePath)!;
    if (!await File(stampPath).exists()) {
      _log.info('No secret-store stamp at $stampPath — creating one.');
      try {
        await _createStamp(stampPath);
      } catch (e) {
        _privateKey = null;
        _publicKey = null;
        _setStatus(SecretStoreStatus.mismatch);
        _log.severe('Failed to create stamp: $e');
        return;
      }
      _setStatus(SecretStoreStatus.newStamp);
      _log.info(
        'Secret-store stamp created. This age key is now bound to '
        '${p.basename(workspacePath)}. Restart with a different key and the '
        'server will refuse to start.',
      );
      return;
    }

    final ok = await _verifyStamp(stampPath);
    if (ok) {
      _setStatus(SecretStoreStatus.verified);
      _log.info('Secret-store stamp verified.');
    } else {
      _privateKey = null;
      _publicKey = null;
      _setStatus(SecretStoreStatus.mismatch);
      _log.severe(
        'SECRET-STORE STAMP MISMATCH for ${p.basename(workspacePath)}. '
        'The supplied age key does not decrypt the existing stamp. Restart '
        'with the correct key, or delete $stampPath plus any encrypted '
        '.env.secrets files to start fresh.',
      );
    }
  }

  /// Caller still validates the supplied key against the stored one before
  /// returning secret values over the wire. With sops-on-the-side, this is
  /// just a constant-time-ish equality check against the in-memory key.
  bool verifyHeaderKey(String candidate) {
    final stored = _privateKey;
    if (stored == null) return false;
    final c = candidate.trim();
    if (c.length != stored.length) return false;
    var diff = 0;
    for (var i = 0; i < stored.length; i++) {
      diff |= c.codeUnitAt(i) ^ stored.codeUnitAt(i);
    }
    return diff == 0;
  }

  /// Decrypt the dotenv-formatted, sops-encrypted secrets file at [path].
  Future<Map<String, String>> loadSecrets(String path) async {
    final file = File(path);
    if (!await file.exists()) return <String, String>{};
    final key = _privateKey;
    if (key == null) {
      throw SecretStoreLockedException(
        'Secret store is locked. Restart the server with LIMOUSINE_AGE_KEY set.',
      );
    }
    final result = await _sopsDecrypt(path, key);
    if (result == null) {
      throw SecretStoreLockedException(
        'Failed to decrypt $path. The file may not be sops-encrypted, or it '
        'was encrypted to a different age recipient than the current key.',
      );
    }
    return Env.parseEnvLines(result.split('\n'));
  }

  /// Encrypt [content] (dotenv KV map) and write the sops-armored result to
  /// [path]. Overwrites in place.
  Future<void> saveSecrets(String path, Map<String, String> content) async {
    final pub = _publicKey;
    if (pub == null) {
      throw SecretStoreLockedException(
        'Secret store is locked. Cannot save encrypted secrets.',
      );
    }
    final plain = Env.serialize(content);
    final encrypted = await _sopsEncrypt(plain, pub);
    if (encrypted == null) {
      throw SecretStoreLockedException('sops encrypt failed for $path');
    }
    final file = File(path);
    await file.parent.create(recursive: true);
    await file.writeAsString(encrypted);
  }

  Future<void> dispose() async {
    _privateKey = null;
    _publicKey = null;
    await _changes.close();
  }

  String? get workspaceDir => _workspaceDir;

  // ---------------------------------------------------------------------------
  // sops/age shell-out helpers
  // ---------------------------------------------------------------------------

  Future<String?> _derivePublicKey(String privateKey) async {
    final proc = await Process.start('age-keygen', ['-y']);
    proc.stdin.add(utf8.encode('$privateKey\n'));
    await proc.stdin.close();
    final out = await proc.stdout.transform(utf8.decoder).join();
    final err = await proc.stderr.transform(utf8.decoder).join();
    final code = await proc.exitCode;
    if (code != 0) {
      _log.warning('age-keygen -y failed: $err');
      return null;
    }
    final pub = out.trim();
    if (!pub.startsWith('age1')) {
      _log.warning('age-keygen -y produced unexpected output: $pub');
      return null;
    }
    return pub;
  }

  /// `sops -d --input-type dotenv --output-type dotenv <path>`. Returns
  /// decrypted dotenv text, or null on any failure.
  Future<String?> _sopsDecrypt(String path, String privateKey) async {
    final env = {
      ...Platform.environment,
      'SOPS_AGE_KEY': privateKey,
    };
    final proc = await Process.start(
      'sops',
      [
        '-d',
        '--input-type', 'dotenv',
        '--output-type', 'dotenv',
        path,
      ],
      environment: env,
      includeParentEnvironment: false,
    );
    final out = await proc.stdout.transform(utf8.decoder).join();
    final err = await proc.stderr.transform(utf8.decoder).join();
    final code = await proc.exitCode;
    if (code != 0) {
      _log.warning('sops -d $path failed (exit $code): $err');
      return null;
    }
    return out;
  }

  /// `sops -e --age <pub> --input-type dotenv --output-type dotenv /dev/stdin`.
  /// Returns encrypted text, or null on failure.
  Future<String?> _sopsEncrypt(String plaintext, String publicKey) async {
    // sops doesn't read /dev/stdin reliably for typed inputs; round-trip via a
    // tempfile to make the input-type detection deterministic.
    final tmp = await Directory.systemTemp.createTemp('limousine-sops-');
    final inputPath = p.join(tmp.path, 'in.env');
    try {
      await File(inputPath).writeAsString(plaintext);
      final proc = await Process.start(
        'sops',
        [
          '-e',
          '--age', publicKey,
          '--input-type', 'dotenv',
          '--output-type', 'dotenv',
          inputPath,
        ],
        includeParentEnvironment: true,
      );
      final out = await proc.stdout.transform(utf8.decoder).join();
      final err = await proc.stderr.transform(utf8.decoder).join();
      final code = await proc.exitCode;
      if (code != 0) {
        _log.warning('sops -e failed (exit $code): $err');
        return null;
      }
      return out;
    } finally {
      try {
        await tmp.delete(recursive: true);
      } catch (_) {}
    }
  }

  Future<void> _createStamp(String stampPath) async {
    final pub = _publicKey!;
    final encrypted = await _sopsEncrypt(
      'LIMOUSINE_STAMP=$_stampPlaintext\n',
      pub,
    );
    if (encrypted == null) {
      throw StateError('sops encrypt failed creating stamp');
    }
    final file = File(stampPath);
    await file.parent.create(recursive: true);
    await file.writeAsString(encrypted);
  }

  Future<bool> _verifyStamp(String stampPath) async {
    final key = _privateKey!;
    final decrypted = await _sopsDecrypt(stampPath, key);
    if (decrypted == null) return false;
    return decrypted.contains('LIMOUSINE_STAMP=$_stampPlaintext');
  }
}

File: limousine/backend/secret_store.py
"""In-memory secrets gateway, backed by `sops` + `age`.
Port of lib/server/secret_store.dart.

Holds the dev's age private key (pasted at startup via the wrapper) in memory;
never touches disk with it. Public key is derived once and cached.

All encryption / decryption is delegated to the `sops` CLI. Limousine does NOT
inject secrets into spawned services — the start command in each project's
limousine.proj does `sops exec-env secrets.env '<cmd>'` itself. This module's
runtime responsibilities are:
  - validate the key on startup against a `secret-store-stamp` (sops-encrypted)
  - serve the secrets editor UI (decrypt for view, encrypt on save)
  - expose the private key so the service manager can forward SOPS_AGE_KEY into
    spawned child processes

Deliberately not called "vault" — that name belongs to HashiCorp Vault, and
this is a much smaller, single-key local-disk feature.
"""

from __future__ import annotations

import asyncio
import enum
import hmac
import logging
import os
import tempfile
from pathlib import Path

from . import env as env_mod

_log = logging.getLogger("limousine.secret_store")


class SecretStoreStatus(enum.Enum):
    # No age private key was supplied at startup. Server runs but services that
    # need encrypted secrets cannot start (their commands will sops-fail).
    MISSING_KEY = "missingKey"

    # Key supplied + stamp existed + decrypted successfully.
    VERIFIED = "verified"

    # Key supplied + no stamp on disk → just created one with this key.
    NEW_STAMP = "newStamp"

    # Key supplied + stamp existed + sops decryption failed. The caller treats
    # this as fatal and exits.
    MISMATCH = "mismatch"


class SecretStoreLockedException(Exception):
    pass


_STAMP_PLAINTEXT = "limousine-secret-store-v2"


class SecretStore:
    def __init__(self) -> None:
        self._private_key: str | None = None  # AGE-SECRET-KEY-…
        self._public_key: str | None = None  # age1…
        self._status: SecretStoreStatus = SecretStoreStatus.MISSING_KEY
        self._workspace_dir: str | None = None
        self.on_status = None  # callback(SecretStoreStatus) — replaces the Dart stream

    @property
    def status(self) -> SecretStoreStatus:
        return self._status

    @property
    def unlocked(self) -> bool:
        return self._private_key is not None

    @property
    def private_key(self) -> str | None:
        """The age private key, exposed so the service manager can forward it
        into child processes as SOPS_AGE_KEY. None if the store is locked."""
        return self._private_key

    @property
    def workspace_dir(self) -> str | None:
        return self._workspace_dir

    @staticmethod
    def stamp_path_for(workspace_path: str | Path) -> str:
        return str(Path(workspace_path).parent / ".limousine" / "secret-store-stamp")

    def _set_status(self, s: SecretStoreStatus) -> None:
        self._status = s
        if self.on_status:
            self.on_status(s)

    async def init_for_workspace(self, workspace_path: str | Path) -> None:
        """Read `LIMOUSINE_AGE_KEY` from the environment, validate it by deriving
        the public key, then verify or create the workspace stamp file."""
        self._workspace_dir = str(Path(workspace_path).parent)
        key_env = os.environ.get("LIMOUSINE_AGE_KEY")
        self._private_key = None
        self._public_key = None

        if not key_env:
            self._set_status(SecretStoreStatus.MISSING_KEY)
            _log.warning(
                "LIMOUSINE_AGE_KEY not set — secret store locked. "
                "Services with encrypted secrets will fail to decrypt."
            )
            return

        pub = await self._derive_public_key(key_env.strip())
        if pub is None:
            self._set_status(SecretStoreStatus.MISMATCH)
            _log.error(
                "LIMOUSINE_AGE_KEY does not parse as a valid age private key. "
                "Restart with a key produced by `age-keygen`."
            )
            return
        self._private_key = key_env.strip()
        self._public_key = pub

        stamp_path = self.stamp_path_for(workspace_path)
        if not Path(stamp_path).exists():
            _log.info("No secret-store stamp at %s — creating one.", stamp_path)
            try:
                await self._create_stamp(stamp_path)
            except Exception as e:
                self._private_key = None
                self._public_key = None
                self._set_status(SecretStoreStatus.MISMATCH)
                _log.error("Failed to create stamp: %s", e)
                return
            self._set_status(SecretStoreStatus.NEW_STAMP)
            _log.info(
                "Secret-store stamp created. This age key is now bound to %s. "
                "Restart with a different key and the server will refuse to start.",
                Path(workspace_path).name,
            )
            return

        ok = await self._verify_stamp(stamp_path)
        if ok:
            self._set_status(SecretStoreStatus.VERIFIED)
            _log.info("Secret-store stamp verified.")
        else:
            self._private_key = None
            self._public_key = None
            self._set_status(SecretStoreStatus.MISMATCH)
            _log.error(
                "SECRET-STORE STAMP MISMATCH for %s. The supplied age key does "
                "not decrypt the existing stamp. Restart with the correct key, "
                "or delete %s plus any encrypted secrets.env files to start fresh.",
                Path(workspace_path).name,
                stamp_path,
            )

    async def unlock(self, private_key: str) -> SecretStoreStatus:
        """Supply the age key at runtime (secrets editor) when it wasn't in the
        env at startup. Derives the pubkey and verifies/creates the stamp, same
        contract as init_for_workspace."""
        if self._workspace_dir is None:
            raise RuntimeError("no workspace open")
        pub = await self._derive_public_key(private_key.strip())
        if pub is None:
            return SecretStoreStatus.MISMATCH
        self._private_key = private_key.strip()
        self._public_key = pub
        stamp_path = str(Path(self._workspace_dir) / ".limousine" / "secret-store-stamp")
        if not Path(stamp_path).exists():
            await self._create_stamp(stamp_path)
            self._set_status(SecretStoreStatus.NEW_STAMP)
        elif await self._verify_stamp(stamp_path):
            self._set_status(SecretStoreStatus.VERIFIED)
        else:
            self._private_key = None
            self._public_key = None
            self._set_status(SecretStoreStatus.MISMATCH)
        return self._status

    def verify_header_key(self, candidate: str) -> bool:
        """Constant-time equality check of an API auth header against the
        in-memory key, before returning secret values over the wire."""
        stored = self._private_key
        if stored is None:
            return False
        return hmac.compare_digest(candidate.strip(), stored)

    async def decrypt_secrets(self, path: str | Path) -> dict[str, str]:
        """Decrypt the dotenv-formatted, sops-encrypted secrets file at [path]."""
        if not Path(path).exists():
            return {}
        key = self._private_key
        if key is None:
            raise SecretStoreLockedException(
                "Secret store is locked. Restart the server with LIMOUSINE_AGE_KEY set."
            )
        result = await self._sops_decrypt(str(path), key)
        if result is None:
            raise SecretStoreLockedException(
                f"Failed to decrypt {path}. The file may not be sops-encrypted, or "
                "it was encrypted to a different age recipient than the current key."
            )
        return env_mod.parse_env_lines(result.split("\n"))

    async def save_secrets(self, path: str | Path, content: dict[str, str]) -> None:
        """Encrypt [content] (dotenv KV map) and write the sops-armored result
        to [path]. Overwrites in place."""
        pub = self._public_key
        if pub is None:
            raise SecretStoreLockedException(
                "Secret store is locked. Cannot save encrypted secrets."
            )
        plain = env_mod.serialize(content)
        encrypted = await self._sops_encrypt(plain, pub)
        if encrypted is None:
            raise SecretStoreLockedException(f"sops encrypt failed for {path}")
        p = Path(path)
        p.parent.mkdir(parents=True, exist_ok=True)
        p.write_text(encrypted)

    def dispose(self) -> None:
        self._private_key = None
        self._public_key = None

    # ------------------------------------------------------------------------
    # sops/age shell-out helpers
    # ------------------------------------------------------------------------

    async def _derive_public_key(self, private_key: str) -> str | None:
        proc = await asyncio.create_subprocess_exec(
            "age-keygen",
            "-y",
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        out_b, err_b = await proc.communicate(f"{private_key}\n".encode())
        if proc.returncode != 0:
            _log.warning("age-keygen -y failed: %s", err_b.decode())
            return None
        pub = out_b.decode().strip()
        if not pub.startswith("age1"):
            _log.warning("age-keygen -y produced unexpected output: %s", pub)
            return None
        return pub

    async def _sops_decrypt(self, path: str, private_key: str) -> str | None:
        """`sops -d --input-type dotenv --output-type dotenv <path>`. Returns
        decrypted dotenv text, or None on any failure."""
        env = {**os.environ, "SOPS_AGE_KEY": private_key}
        proc = await asyncio.create_subprocess_exec(
            "sops",
            "-d",
            "--input-type",
            "dotenv",
            "--output-type",
            "dotenv",
            path,
            env=env,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        out_b, err_b = await proc.communicate()
        if proc.returncode != 0:
            _log.warning("sops -d %s failed (exit %s): %s", path, proc.returncode, err_b.decode())
            return None
        return out_b.decode()

    async def _sops_encrypt(self, plaintext: str, public_key: str) -> str | None:
        """`sops -e --age <pub> --input-type dotenv --output-type dotenv <file>`.
        Returns encrypted text, or None on failure."""
        # sops doesn't read /dev/stdin reliably for typed inputs; round-trip via
        # a tempfile (with a .env extension) to make input-type detection
        # deterministic.
        with tempfile.TemporaryDirectory(prefix="limousine-sops-") as tmp:
            input_path = str(Path(tmp) / "in.env")
            Path(input_path).write_text(plaintext)
            proc = await asyncio.create_subprocess_exec(
                "sops",
                "-e",
                "--age",
                public_key,
                "--input-type",
                "dotenv",
                "--output-type",
                "dotenv",
                input_path,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            out_b, err_b = await proc.communicate()
            if proc.returncode != 0:
                _log.warning("sops -e failed (exit %s): %s", proc.returncode, err_b.decode())
                return None
            return out_b.decode()

    async def _create_stamp(self, stamp_path: str) -> None:
        assert self._public_key is not None
        encrypted = await self._sops_encrypt(
            f"LIMOUSINE_STAMP={_STAMP_PLAINTEXT}\n",
            self._public_key,
        )
        if encrypted is None:
            raise RuntimeError("sops encrypt failed creating stamp")
        p = Path(stamp_path)
        p.parent.mkdir(parents=True, exist_ok=True)
        p.write_text(encrypted)

    async def _verify_stamp(self, stamp_path: str) -> bool:
        assert self._private_key is not None
        decrypted = await self._sops_decrypt(stamp_path, self._private_key)
        if decrypted is None:
            return False
        return f"LIMOUSINE_STAMP={_STAMP_PLAINTEXT}" in decrypted
